platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/migratable/FeatureProcessor.java

```diff
@@ -106,35 +117,145 @@ public Feature getFeature(MigrationReport report, String id) {
+   * @param region the region where to install the features
+   * @param jfeatures the features to install
+   * @return <code>true</code> if the features were installed successfully; <code>false</code>
+   *     otherwise
+   */
+  public boolean installFeatures(
+      ProfileMigrationReport report, String region, Set<JsonFeature> jfeatures) {
+    final Set<String> ids = jfeatures.stream().map(JsonFeature::getId).collect(Collectors.toSet());
+
     return run(
         report,
-        feature,
+        region,
+        ids.stream(),
         Operation.INSTALL,
-        id -> service.installFeature(id, FeatureProcessor.NO_AUTO_REFRESH));
+        () -> service.installFeatures(ids, region, FeatureProcessor.NO_AUTO_REFRESH));
   }
 
   /**
-   * Uninstalls the specified feature.
+   * Uninstalls the specified set of features.
    *
-   * @param report the report where to record errors if unable to uninstall the feature
-   * @param feature the feature to uninstall
-   * @return <code>true</code> if the feature was uninstalled successfully; <code>false</code>
+   * @param report the report where to record errors if unable to uninstall the features
+   * @param features the features to install keyed by the region they should be uninstalled in
+   * @return <code>true</code> if the features were uninstalled successfully; <code>false</code>
    *     otherwise
    */
-  public boolean uninstallFeature(ProfileMigrationReport report, Feature feature) {
+  public boolean uninstallFeatures(
+      ProfileMigrationReport report, Map<String, Set<Feature>> features) {
+    return features
+        .entrySet()
+        .stream()
+        .allMatch(e -> uninstallFeatures(report, e.getKey(), e.getValue()));
+  }
+
+  /**
+   * Uninstalls the specified set of features in the specified region.
+   *
+   * @param report the report where to record errors if unable to uninstall the features
+   * @param region the region where to uninstall the features
+   * @param features the features to uninstall
+   * @return <code>true</code> if the features were uninstalled successfully; <code>false</code>
+   *     otherwise
+   */
+  public boolean uninstallFeatures(
+      ProfileMigrationReport report, String region, Set<Feature> features) {
+    // ----------------------------------------------------------------------------------------
+    // to work around a bug in Karaf where for whatever reasons, it checks if a feature is
+    // required to determine if it is installed and therefore can be uninstalled, we will first
+    // go through and mark them all required
+    //
+    // see Karaf class: org.apache.karaf.features.internal.service.FeaturesServiceImpl.java
+    // in the uninstallFeature() method where it checks the requirements table and if not found
+    // in there, it later reports it is not installed even though it is actually installed
+    //
+    // Update: turns out that what Karaf means by install is "require this feature" and by uninstall
```

Comment:
I prefer keeping them as they provide history as to why the design was done like this. The first one describes the workaround still being done because of the weird behavior (explained in the second paragraph) in Karaf which even after the email exchange with them was not fully explained.